File: backend/app/models/course.py
from datetime import datetime, timezone

from sqlalchemy import Column, Integer, String, Text, DateTime
from ..database import Base


class Course(Base):
    __tablename__ = "courses"

    id = Column(Integer, primary_key=True, autoincrement=True)
    title = Column(String(200), nullable=False)
    slug = Column(String(100), unique=True, nullable=False)
    description = Column(Text, default="")
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime, default=lambda: datetime.now(timezone.utc),
                        onupdate=lambda: datetime.now(timezone.utc))


class Chapter(Base):
    __tablename__ = "chapters"

    id = Column(Integer, primary_key=True, autoincrement=True)
    course_id = Column(Integer, nullable=False)
    title = Column(String(200), nullable=False)
    order_index = Column(Integer, nullable=False)


class Lesson(Base):
    __tablename__ = "lessons"

    id = Column(Integer, primary_key=True, autoincrement=True)
    chapter_id = Column(Integer, nullable=False)
    title = Column(String(200), nullable=False)
    content_md = Column(Text, default="")
    order_index = Column(Integer, nullable=False)
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime, default=lambda: datetime.now(timezone.utc),
                        onupdate=lambda: datetime.now(timezone.utc))


class Quiz(Base):
    __tablename__ = "quizzes"

    id = Column(Integer, primary_key=True, autoincrement=True)
    lesson_id = Column(Integer, nullable=False)
    title = Column(String(200), default="")


class QuizQuestion(Base):
    __tablename__ = "quiz_questions"

    id = Column(Integer, primary_key=True, autoincrement=True)
    quiz_id = Column(Integer, nullable=False)
    question_text = Column(Text, nullable=False)
    options = Column(Text, nullable=False)  # JSON array
    correct_answer_index = Column(Integer, nullable=False)
    explanation = Column(Text, default="")

File: src/__tests__/factories/course.ts
import type {
  CourseSummary,
  CourseDetail,
  LessonDetail,
  Quiz,
  ChapterDetail,
  QuizQuestion,
} from '../../types'

export function createCourse(overrides?: Partial<CourseSummary>): CourseSummary {
  return {
    id: 1,
    title: 'Test Course',
    slug: 'test-course',
    description: 'A test course',
    difficulty: 'beginner',
    duration_minutes: 0,
    tags: [],
    ...overrides,
  }
}

export function createCourseDetail(overrides?: Partial<CourseDetail>): CourseDetail {
  return {
    ...createCourse(),
    chapters: [createChapter()],
    ...overrides,
  }
}

export function createChapter(overrides?: Partial<ChapterDetail>): ChapterDetail {
  return {
    id: 1,
    title: 'Test Chapter',
    order_index: 1,
    lessons: [createLesson()],
    ...overrides,
  }
}

export function createLesson(overrides?: Partial<LessonDetail>): LessonDetail {
  return {
    id: 1,
    chapter_id: 1,
    title: 'Test Lesson',
    content_md: '# Test\n\nLesson content',
    order_index: 1,
    duration_minutes: 0,
    ...overrides,
  }
}

export function createQuiz(overrides?: Partial<Quiz>): Quiz {
  return {
    id: 1,
    lesson_id: 1,
    title: 'Test Quiz',
    questions: [createQuizQuestion()],
    ...overrides,
  }
}

export function createQuizQuestion(overrides?: Partial<QuizQuestion>): QuizQuestion {
  return {
    id: 1,
    question_text: 'What is AI?',
    options: ['A', 'B', 'C', 'D'],
    explanation: 'A is correct',
    ...overrides,
  }
}
